### PR TITLE
Implement instant search screen - 3

### PR DESCRIPTION
**Pull request by msasikanth**
https://app.clubhouse.io/simpledotorg/story/2029/implement-instant-search-screen

### CODE REVIEW

**Review by Honey14 — CHANGES_REQUESTED** [review]
Looks great overall. Left few comments. Please take a look.

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchEffectHandler.kt [hunk 1]
```diff
@@ -18,8 +18,18 @@ class InstantSearchEffectHandler @Inject constructor(
...
       .addTransformer(LoadCurrentFacility::class.java, loadCurrentFacility())
       .addTransformer(LoadAllPatients::class.java, loadAllPatients())
+      .addTransformer(SearchWithCriteria::class.java, searchWithCriteria())
       .build()
 
+  private fun searchWithCriteria(): ObservableTransformer<SearchWithCriteria, InstantSearchEvent> {
+    return ObservableTransformer { effects ->
+      effects
+          .observeOn(schedulers.io())
+          .map { patientRepository.search2(it.criteria, it.facility.uuid) }
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchEffectHandler.kt:51]: We will be renaming this to `search` once we remove older search implementation. So for now this should be fine.

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package org.simple.clinic.instantsearch
+
+import org.simple.clinic.facility.Facility
+
+data class InstantSearchModel(
+    val facility: Facility?
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt]: When the screen is first created we won't have any `Facility` in the model. This is the way for us to identify whether or not we have a facility and act accordingly upon that. For example.

```kt
    if (model.hasFacility)
      effects.add(ValidateSearchQuery(model.searchQuery.orEmpty()))
    else
      effects.add(LoadCurrentFacility)
```

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchUpdate.kt [hunk 3]
```diff
@@ -3,15 +3,22 @@ package org.simple.clinic.instantsearch
...
     return when (event) {
       is CurrentFacilityLoaded -> next(model.facilityLoaded(event.facility), LoadAllPatients(event.facility))
-      is AllPatientsLoaded -> noChange()
+      is AllPatientsLoaded -> allPatientsLoaded(model, event)
       is SearchResultsLoaded -> noChange()
     }
   }
+
+  private fun allPatientsLoaded(model: InstantSearchModel, event: AllPatientsLoaded): Next<InstantSearchModel, InstantSearchEffect> {
+    if (model.hasSearchQuery) return noChange()
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchUpdate.kt:55]: WDYM? when there is a search query we return early without any changes.
```kt
return noChange()
```

We only continue ahead if the query is empty or null.

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchUpdate.kt [hunk 4]
```diff
@@ -12,10 +12,16 @@ class InstantSearchUpdate : Update<InstantSearchModel, InstantSearchEvent, Insta
...
       is AllPatientsLoaded -> allPatientsLoaded(model, event)
-      is SearchResultsLoaded -> noChange()
+      is SearchResultsLoaded -> searchResultsLoaded(model, event)
     }
   }
 
+  private fun searchResultsLoaded(model: InstantSearchModel, event: SearchResultsLoaded): Next<InstantSearchModel, InstantSearchEffect> {
+    if (!model.hasSearchQuery) return noChange()
+
+    return dispatch(ShowPatientSearchResults(event.patientsSearchResults, model.facility!!))
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchUpdate.kt:51]: I hope this address that concern?
https://github.com/simpledotorg/simple-android/pull/2251#discussion_r551309790

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt [hunk 5]
```diff
@@ -20,4 +20,8 @@ data class InstantSearchModel(
   fun facilityLoaded(facility: Facility): InstantSearchModel {
     return copy(facility = facility)
   }
+
+  fun searchQueryChanged(searchQuery: String): InstantSearchModel {
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt:30]: Ah yes, that will be done in later commits. This method was implemented here for the test purpose.

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt [hunk 6]
```diff
@@ -1,10 +1,12 @@
 package org.simple.clinic.instantsearch
 
 import org.simple.clinic.facility.Facility
+import org.simple.clinic.patient.businessid.Identifier
 
 data class InstantSearchModel(
     val facility: Facility?,
-    val searchQuery: String?
+    val searchQuery: String?,
+    val additionalIdentifier: Identifier?
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt:9]: The identifier contains the BP passport information when a user scans a BP passport we need to pass the identifier to patient summary or create page.

**Review by Honey14 — APPROVED** [review]
Looks good overall

**Pull request merged**